streams/src/test/java/org/apache/kafka/streams/integration/StreamsUpgradeTestIntegrationTest.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.integration;
+
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.integration.utils.EmbeddedKafkaCluster;
+import org.apache.kafka.streams.integration.utils.IntegrationTestUtils;
+import org.apache.kafka.streams.processor.internals.assignment.SubscriptionInfo;
+import org.apache.kafka.streams.tests.StreamsUpgradeTest;
+import org.apache.kafka.test.IntegrationTest;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.time.Duration;
+import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.apache.kafka.common.utils.Utils.mkProperties;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@Category(IntegrationTest.class)
+public class StreamsUpgradeTestIntegrationTest {
```

Comment:
I accidentally broke the version probing system test when I made the code changes for this PR, and it was super-hard to track down why in system test form, so I added an integration test for the system test code. This seems like a reasonable practice, since we don't seem to be able to run the system tests locally; an integration test gives us the ability to find out right away if we have broken the system test logic, and also gives us a way to verify a fix without waiting a minimum of a half-hour for the system test to run.

Reply:
I think this is ok.   But let's do it on a case by case basis, i.e., not require a likewise integration test for each system test.  I'm don't think that's your intent, but I just wanted to be clear.

What was the change that broke the system test?

Reply:
Sounds wise :)

The FutureVersion stuff in the upgrade test previously relied on serialization methods in SubscriptionInfo that now no longer exist. I tried just deleting all the extra fields and just encoding the used and supported versions only in the message, reasoning that these are all that's really required to test version probing.

Well, it turns out the test does a little more than I thought, and it actually expects the whole cluster to upgrade to the "future version", so it actually does need to encode the tasks etc.

What I've done now is just inline all the logic to encode the full "future version" subscription the same way we always have, so that the test can continue to work the same way. I think this is reasonable, since the logic of the test is explicitly mocking out a "future" message, not intentionally depending on anything about the way that the current messages are encoded.